File: Languages/go/algorithms-dsa/leetcode/dynamic-programming/partition_equal_subset_sum.go
package main

import "fmt"

/*
  LC 416 - Partition Equal Subset Sum
  Topic: Dynamic Programming (0/1 Knapsack)
  Difficulty: Medium
  Time: O(n*sum) | Space: O(sum)
*/

func canPartition(nums []int) bool {
	total := 0
	for _, n := range nums {
		total += n
	}
	if total%2 != 0 {
		return false
	}
	target := total / 2
	dp := make([]bool, target+1)
	dp[0] = true
	for _, n := range nums {
		for j := target; j >= n; j-- {
			dp[j] = dp[j] || dp[j-n]
		}
	}
	return dp[target]
}

func main() {
	if canPartition([]int{1, 5, 11, 5}) {
		fmt.Println("PASS: canPartition [1,5,11,5]")
	} else {
		fmt.Println("FAIL: canPartition [1,5,11,5]")
	}

	if !canPartition([]int{1, 2, 3, 5}) {
		fmt.Println("PASS: canPartition [1,2,3,5]")
	} else {
		fmt.Println("FAIL: canPartition [1,2,3,5]")
	}
}

File: Languages/python/algorithms-dsa/leetcode/dynamic-programming/partition_equal_subset_sum.py
"""
LeetCode #416 - Partition Equal Subset Sum
Topic: Dynamic Programming
Difficulty: Medium

Determine if array can be partitioned into two equal-sum subsets using subset sum DP.

Time Complexity: O(n * target)
Space Complexity: O(target)
"""


def can_partition(nums: list[int]) -> bool:
    total = sum(nums)
    if total % 2 != 0:
        return False
    target = total // 2
    dp = [False] * (target + 1)
    dp[0] = True
    for num in nums:
        for j in range(target, num - 1, -1):
            dp[j] = dp[j] or dp[j - num]
    return dp[target]


if __name__ == "__main__":
    assert can_partition([1, 5, 11, 5]) is True
    assert can_partition([1, 2, 3, 5]) is False
    assert can_partition([2, 2]) is True
    assert can_partition([1]) is False
    print("All tests passed!")
